### DISCUSSION TITLE
importing sionna gives DRJIT_LIBLLVM_PATH error on M2 Max

### DISCUSSION THREAD

**Discussion by galfredo-apple** [NVlabs/sionna#179] [Q&A] [answered]
When I do <import sionna> I get 

AttributeError: jit_init_thread_state(): the LLVM backend is inactive because the LLVM shared library ("libLLVM.dylib") could not be found! Set the DRJIT_LIBLLVM_PATH environment variable to specify its path.

I am running a M2 Max and I have done 
1. brew install llvm and it was successful 
2.pip install tensorflow-macos

and the environment variable is set to DRJIT_LIBLLVM_PATH=/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/libclang.dylib

Thanks for any help.

**Comment by galfredo-apple** [comment]
I figured out that you have to import tensorflow , then drjit and then you can import sionna

but then when I try to load_scene of an example scene I get the same error message.

 I have mitsuba3 confirmed downloaded and DRJIT_LIBLLVM_PATH=/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/libclang.dylib

**Comment by merlinND** [comment]
Hello @galfredo-apple,

On macOS with LLVM installed via Homebrew, LLVM should be found automatically, so you shouldn't need to set the `DRJIT_LIBLLVM_PATH` environment variable.

If for some reason it is not (perhaps due to installing the x86 version of LLVM instead of the native one, or vice versa?), then `DRJIT_LIBLLVM_PATH` should point to `libLLVM.dylib`, as indicated by the error message. Currently you have it pointing to `libclang.dylib`.

**Reply by galfredo-apple** [reply]
Hmm yeah I have it pointing now to DRJIT_LIBLLVM_PATH=/usr/local/opt/llvm/lib/libLLVM.dylib but I get the same error message when I try to run 

scenic = sionna.rt.scene.load_scene( sionna.rt.scene.munich ) 

I used brew install llvm so it should be right one...

**Reply by merlinND** [reply]
Just to check, do you see the correct value when printing `os.environ['DRJIT_LIBLLVM_PATH']` in your Python script?

Otherwise, my best guess is that the Python you're running is native but the LLVM version installed is x86, or the other way around. This answer helps check which architecture(s) a library is built for: https://stackoverflow.com/questions/1085137/how-do-i-determine-the-target-architecture-of-static-library-a-on-mac-os-x

**Comment by galfredo-apple** [comment]
I was able to get the ray tracing tool to work but it has to be on an x84 and on ubuntu. x84 w/ WSL or arm64 seems to be buggy. I hope one day I can run it natively on the m-chip and maybe even use the m-chip gpus :)

**Comment by kpasad** [comment]
I saw the same issue, but thanks to the tip on importing drjit , i was able to import sionna on Azure VM (ubuntu) with a conda virtual environment. However, the error persist when i use VScode with remote SSH. 
For ubuntu, what would be the DRJIT_LIBLLVM_PATH ? There is no libllvm.so .

**Reply by galfredo-apple** [reply]
Same, for some reason I was getting that error when I run it with Jupyter notebook in VScode but I'm able to run things off the terminal. Haven't been able to figure out why yet.

**Reply by merlinND** [reply, marked as answer]
Hello @kpasad,

This sounds like a search path issue. In Python, you can always check the current search path in your script or notebook by printing `os.environ['DRJIT_LIBLLVM_PATH'], os.environ['PATH'], os.environ['LD_LIBRARY_PATH']`.

> There is no libllvm.so

If there is no libllvm on your system, you need to install llvm-dev, e.g. with `apt`.

**Reply by kpasad** [reply]
the libllvm.so is missing. As you said @merlinND ,  I did a apt-get  install and now i can run from VSCode.

**Comment by nbecker** [comment]
Running on Fedora F38.  I've been using sionna for months now, then this
morning I got
AttributeError: jit_init_thread_state(): the LLVM backend is inactive
because the LLVM shared library ("libLLVM.so") could not be found! Set the
DRJIT_LIBLLVM_PATH environment variable to specify its path.

I found /usr/lib64/libLLVM.so was missing, and I did
sudo dnf install llvm-devel which fixed the problem.

Very strange since it was working for months before.

On Tue, Aug 8, 2023 at 10:15 AM Merlin Nimier-David <
***@***.***> wrote:

> Closed #179 <https://github.com/NVlabs/sionna/discussions/179> as
> resolved.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NVlabs/sionna/discussions/179>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAA4VUCUFSS2YLW2QKQSCWDXUJCYVANCNFSM6AAAAAA2YHLOY4>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***
> com>
>


-- 
*Those who don't understand recursion are doomed to repeat it*

**Reply by merlinND** [reply]
Hello @nbecker,

Sionna tries to use the CUDA backend first, and falls back on LLVM if it doesn't detect supported GPUs.
Is it possible that something changed / broke on your system, that lead Sionna to fallback to the LLVM backend for the first time? (and therefore requiring the library to be installed)

**Comment by nbecker** [comment]
On Tue, Aug 15, 2023 at 7:53 AM Merlin Nimier-David <
***@***.***> wrote:

> Hello @nbecker <https://github.com/nbecker>,
>
> Sionna tries to use the CUDA backend first, and falls back on LLVM if it
> doesn't detect supported GPUs.
> Is it possible that something changed / broke on your system, that lead
> Sionna to fallback to the LLVM backend for the first time? (and therefore
> requiring the library to be installed)
>
You are exactly correct, kernel update caused library/driver mismatch so
cuda was not loaded.  Thanks!
